feat: pqk + quotas cleanup

Van0k opened the pull request.

Van0k committed:
fix: pqk + quotas cleanup

0xmikko committed:
fix: gas saving optimisation

0xmikko committed:
fix: memory -> calldata

0xmikko committed:
fix: memory -> calldata

Van0k committed:
feat: pqk tests

Van0k committed:
chore: exceptions cleanup

Van0k committed:
feat: update quota events moved to pqk

lekhovitsky commented on contracts/credit/CreditFacadeV3.sol:
```diff
@@ -958,8 +958,6 @@ contract CreditFacadeV3 is ICreditFacadeV3, ACLNonReentrantTrait {
             maxQuota: uint96(Math.min(type(uint96).max, maxQuotaMultiplier * debtLimits.maxDebt))
         }); // U:[FA-34]
-
-        emit UpdateQuota({creditAccount: creditAccount, token: token, quotaChange: realQuotaChange}); // U:[FA-34]
```
probably need to remove the `realQuotaChange` variable to avoid warning

lekhovitsky commented on contracts/interfaces/IPoolQuotaKeeperV3.sol:
```diff
@@ -19,8 +19,11 @@ struct AccountQuota {
-    /// @dev Emits when CA's quota for token is changed
-    event ChangeAccountQuota(address creditAccount, address token, uint96 oldQuota, uint96 newQuota);
+    /// @dev Emits when a quota for an account is updated
+    event UpdateQuota(address indexed creditAccount, address indexed token, int96 realQuotaChange);
```
maybe just keep it as `quotaChange` in the event?

lekhovitsky commented on contracts/interfaces/IPoolQuotaKeeperV3.sol:
```diff
@@ -19,8 +19,11 @@ struct AccountQuota {
+    event UpdateQuota(address indexed creditAccount, address indexed token, int96 realQuotaChange);
+
+    /// @dev Emits when a quota for an account is updated
+    event RemoveQuota(address indexed creditAccount, address indexed token);
```
is it actually needed? i.e., is removing quota functionally different from updating it with an appropriate `quotaChange` value?

Van0k committed:
Merge branch 'main' into pqk-cleanup

Van0k committed:
feat: better quota revenue update on pqk update

Van0k committed:
feat: controller timelock missing functions + improvements

Van0k commented:
Changes:
- `updateQuota` is now emitted in PQK, `removeQuota` event added
- Pool now updates quota revenue to revenue taken from new PQK on PQK update
- Added missing functions to ControllerTimelock: `setTokenLimit`, `setTokenQuotaIncreaseFee`, `setTotalDebtLimit`, `setWithdrawFee`, `setMinQuotaRate`, `setMaxQuotaRate`
- Split `setDebtLimits` into `setMinDebtLimit` and `setMaxDebtLimit`
- Updated `setCreditManagerDebtLimit` to work with total debt being tracked in CF

lekhovitsky committed:
feat: move staking and gauge to governance

lekhovitsky committed:
feat: move wm, aa and bl to core

lekhovitsky commented on contracts/interfaces/IExceptions.sol:
```diff
@@ -3,87 +3,128 @@
-/// @dev Thrown if an access-restricted function is called by an address that is not
-///      the connected Credit Facade
-error CallerNotCreditFacadeException();
+/// @dev Thrown on configurator attempting to add more than 256 collateral tokens
```
should be more than 255

lekhovitsky commented on contracts/libraries/QuotasLogic.sol:
```diff
@@ -76,76 +37,43 @@ library QuotasLogic {
+    /// @param rate Rate for current token
+    /// @param change Real change in quota
+    function calcQuotaRevenueChange(uint16 rate, int256 change) internal pure returns (int256) {
+        return int256(change) * int256(uint256(rate)) / int16(PERCENTAGE_FACTOR);
```
no need to cast change to `int256`

lekhovitsky commented on contracts/pool/PoolQuotaKeeperV3.sol:
```diff
@@ -192,7 +230,7 @@ contract PoolQuotaKeeperV3 is IPoolQuotaKeeperV3, ACLNonReentrantTrait, Contract
         for (uint256 i; i < len;) {
             address token = tokens[i];
-            if (token == address(0)) break;
+            if (token == address(0)) break; // U: [PQK-16]
```
we probably no longer need this check here

lekhovitsky commented on contracts/pool/PoolQuotaKeeperV3.sol:
```diff
@@ -202,25 +240,32 @@ contract PoolQuotaKeeperV3 is IPoolQuotaKeeperV3, ACLNonReentrantTrait, Contract
                 quoted--;
 
+                uint16 rate = tokenQuotaParams.rate;
+                uint96 totalQuoted = tokenQuotaParams.totalQuoted;
```
can avoid reading it here and simply do `tokenQuotaParams.totalQuoted -= quoted;` on L250

lekhovitsky commented on contracts/pool/PoolQuotaKeeperV3.sol:
```diff
@@ -111,77 +109,117 @@ contract PoolQuotaKeeperV3 is IPoolQuotaKeeperV3, ACLNonReentrantTrait, Contract
-            realQuotaChange = QuotasLogic.calcIncreaseQuotaChange(tokenQuotaParams, quotaChange);
+            // If the requested increase amount is above capacity, only the remaining capacity will
+            // be provided as a quota
+            realQuotaChange = QuotasLogic.calcRealQuotaIncreaseChange(totalQuoted, limit, realQuotaChange); // U: [PQK-15]
```
worth adding a check whether `realQuotaChange` is 0 to avoid unnecessary computation and storage writes

also, if `quoted == 0`, without this check the function would enable token even if `realQuotaChange == 0`

lekhovitsky commented on contracts/pool/PoolQuotaKeeperV3.sol:
```diff
@@ -111,77 +109,117 @@ contract PoolQuotaKeeperV3 is IPoolQuotaKeeperV3, ACLNonReentrantTrait, Contract
             bool enableToken,
             bool disableToken
         )
+    {
```
since we're not emitting event in CF, we no longer need to return `realQuotaChange` here and in the CM

Van0k committed:
fix: small fixes to pqk

lekhovitsky committed:
fix: pool quota keeper improvements

* quotaChange -> requestedQuotaChange
* realQuotaChange -> quotaChange
* don't enable token if newQuoted <= 1

lekhovitsky committed:
fix: rename quotaProfits -> quotaFees

lekhovitsky committed:
fix: rename quotaProfits -> quotaFees (vol. 2)
